api/signal_hook.h

```diff
@@ -0,0 +1,235 @@
+#ifndef BAG_SIGNAL_HOOK
+#define BAG_SIGNAL_HOOK
+
+#include <iostream>
+#include <ostream>
+#include <string>
+
+// from https://man7.org/linux/man-pages/man2/sigaction.2.html
+// and https://man7.org/linux/man-pages/man7/feature_test_macros.7.html
+#ifdef _POSIX_C_SOURCE
+
+#include <string.h>
+#include <signal.h>
+#include <array>
+#include <csignal>
+#include <cstdio>
+#include <unistd.h>
+
+static void (*cleanup_functions[128])(int);
+static int cleanup_functions_count = 0;
+
+// Returns 0 on success, 1 if too many functions have been set
+// This is not thread safe, only call it from a single thread
+int add_cleanup_function(void (*f)(int)) {
+    if (cleanup_functions_count >= 128) {
+        return 1;
+    }
+
+    cleanup_functions[cleanup_functions_count] = f;
+    cleanup_functions_count++;
+
+    return 0;
+}
+
+
+static void bag_handler(int signal);
+
+struct BagAbortHook;
+
+static BagAbortHook* current_hook;
+
+static int bag_string_length(const char* string) {
+    int length = 0;
+    while(true) {
+        if (string[length] == 0) {
+            break;
+        }
+        length++;
+    }
+
+    return length;
+}
+
+struct BagAbortHook {
+    BagAbortHook* previous;
+    //void (*old_handler[255])(int);
+    //void (*current_handler[255])(int);
+    struct sigaction old_handlers[255];
```

Comment:
The size of these handler arrays should be a constant defined by a macro (i.e. #define somewhere).